src/client/datascience/notebook/notebookEditorProvider.ts

```
@@ -166,16 +168,18 @@ export class NotebookEditorProvider implements INotebookEditorProvider {
         let editor = this.notebookEditorsByUri.get(uri.toString());
         if (!editor) {
             const notebookProvider = this.serviceContainer.get<INotebookProvider>(INotebookProvider);
+            const executionService = this.serviceContainer.get<INotebookExecutionService>(INotebookExecutionService);
```

Why do this here instead of as an explicit dependency? These implicit dependencies cause problems in tests and during shutdown.

---

Yes i hate this too, this is because of a cyclic dependency.
I tried to remove these, and for that we need to remove some of the dup events in some of the classes, i have an issue for that (left it for later as it ends up changing more than i'd like).
When VSC NB is stable, then it should be much much easier to make the change,

For now, I'd prefer to leave this.

> during shutdown.

Not sure how this causes issues with shutdowns. I'd like to understand that.

---

During shutdown the service container has unbound everything. If some dispose method causes code to run, and that code happens to go through here, this code will crash. 

---
In reply to: [446306952](https://github.com/microsoft/vscode-python/pull/12570#discussion_r446306952) [](ancestors = 446306952)

---

You can alleviate that problem by using the new tryGet that I added. 

---
In reply to: [446309380](https://github.com/microsoft/vscode-python/pull/12570#discussion_r446309380) [](ancestors = 446309380,446306952)

---

>  If some dispose method causes code to run, and that code happens to go through here, this code will crash.

Isn't that a problem with the style of coding. I.e. when disposing we shoudln't try to get new items from container, we shoudl dispose what we have and have less async stuff in a dispose. 

Cuz it feels like we're merely failing silently using a try.get, when in reality, it should never happen except in test cases where things are tearing down - and that gets down to code doing the wrong things in the dispose/tear down.

I think this is because some of our disposes might get objects (from container) using providers instead of registering the object that needs to be disposed.

---

I.e. `tryGet` is a testing specific code, and feels like a work around instead of not writing code that results in such behavior.

---

No mostly it's objects having complicated dispose patterns that do a lot of work (sometimes making calls to other services to 'shutdown') and those other calls have side effects. Eventually a side effect can end up causing other code to run that didn't expect to run during dispose - example - getActiveInterpreter. This doesn't work during dispose, but it's required to do a bunch of work anywhere which might be necessary in order to figure out somethign during cleanup.

---
In reply to: [446320984](https://github.com/microsoft/vscode-python/pull/12570#discussion_r446320984) [](ancestors = 446320984)

---

tryGet is for behaving gracefully during shutdown. Alternatively you can track shutdown in your class and not do things when shutting down (like opening a notebook).

---
In reply to: [446321178](https://github.com/microsoft/vscode-python/pull/12570#discussion_r446321178) [](ancestors = 446321178)

---

> No mostly it's objects having complicated dispose patterns that do a lot of work

That's my point, shouldn't we try to simplify these instead of continuing with such complicated patterns.

---

Yes and the first step would be to make any new code we write be resilient (or maybe throw a specific error) if it's called during shutdown when it shouldn't be. But it can be hard for a piece of code to know if it's 'allowed' to be called during shutdown. It's safer/easier to just make it behave as much as possible like it should during shutdown.